Banner additions

mejiaj opened the pull request:
Closes #1.

![image](https://github.com/uswds/uswds-next/assets/3385219/2256cc1e-69c5-421d-8e12-1c68eb50ebdb)

## Major changes

- **Code formatting.** Add Prettier for consistent formatting `npm run prettier:js:fix`.
- **Domain variants.** Support for `mil` domains, but `gov` by default.
- **Language variants.** Support for default English and Spanish via `lang` attribute.
- **Customizable content.** Content can be overridden via slots.
- **Visual regression fixes.**
  - Add border box to fix columns.
  - Fix icons.

## Features

### Domain support

#### .Gov Sites

Uses `.gov` by default. Also supports passing it explicitly via `tld="gov"`.

```html
<usa-banner></usa-banner>
<usa-banner tld="gov"></usa-banner>
```

#### .Mil Sites

Added support for `mil` domains via `tld="mil`".

```html
<usa-banner tld="mil"></usa-banner>
```

#### Unsupported domains

Unsupported domains default to `.gov` variant.

```html
<usa-banner tld="biz"></usa-banner>
```

### Language support

Basic support for current English and Spanish variants.

**Default Spanish banner**
```html
<usa-banner lang="es"></usa-banner>
```

#### Unsupported languages

Unsupported languages default to English.

```diff
- <usa-banner lang="yor"></usa-banner>
+ <usa-banner label="Custom Aria label">
+   <span slot="banner-text">Oju opo wẹẹbu osise ti ijọba Amẹrika</span>
+   <span slot="banner-action">Banner action</span>
+   <span slot="domain-heading">Domain info</span>
+   <span slot="domain-text">
+     Lorem ipsum dolor sit amet consectetur adipisicing elit. Hic, libero.
+   </span>
+   <span slot="https-heading">HTTPS info</span>
+   <span slot="https-text">
+     Lorem ipsum dolor sit amet consectetur adipisicing elit. Iusto, exercitationem.
+   </span>
+ </usa-banner>
```


You can add your own by leaving off language and adding your own language via slots:

| Slot name | Default |
|:--------|:--------|
| `banner-text` | An official website of the United States government. |
| `banner-action` | Here’s how you know | 
| `domain-heading` | Official websites use TLD | 
| `domain-text` | A <strong>TLD</strong> website belongs to an official government organization in the United States. | 
| `https-heading` | Secure .TLD websites use HTTPS | 
| `https-text` | A <strong>lock</strong> (<span class="icon-lock">${svgLock}</span>) or <strong>https://</strong> means you’ve safely connected to the .TLD website. Share sensitive information only on official, secure websites. |

Commit:
Add eslint and prettier.

Commit:
Add icons for mobile.

Commit:
Add support for top level domain `mil` and `gov`.

Commit:
Add support for Spanish banner with optional aria label.

Commit:
Use a single template for both English and Spanish version.

- Add default data for English & Spanish.
- Re-use banner action value for Desktop.

Commit:
Update test template.

mejiaj commented on web-components/eslint.config.js:
> [!NOTE]  
> VS Code doesn't pick up this nested config. Consider initializing a `package.json` in root directory.

heymatthenry replied:
Yep. We'll have to have one eslint config in the root of the monorepo.

mejiaj commented on web-components/index.html:
```diff
@@ -16,7 +16,23 @@
+  </span>
+
+</usa-banner>
+
```
![image](https://github.com/uswds/uswds-next/assets/3385219/b5247a53-2272-49cd-9534-a9f47ccfb3e8)

mahoneycm replied:
Cool to see the level of customization we can accomplish with slots and that data object!


I'm curious, any idea what users who have customized the banner further would be able to accomplish? Things like removing the icons, or [Vote.gov](https://vote.gov/)'s example where they added a language selector to their banner

heymatthenry replied:
Now that Storybook is ready to merge (thanks to you, of course, @mejiaj), can the variants go into a Story?

mejiaj replied:
That could be a future enhancement. Right now there's currently no available generic `<slot>` element.

mejiaj commented on web-components/src/components/usa-banner/index.js:
```diff
@@ -20,19 +31,146 @@ export class UsaBanner extends LitElement {
+        },
+        domain: {
+          heading: "Official websites use",
+          text1: "A",
+          text2:
+            "website belongs to an official government organization in the United States.",
```
> [!NOTE]  
> Broken up to better support `.mil` and `.gov` domains. We can re-use these strings without having separate objects for those variants.

Alternatively, we'd need separate objects for Gov/Mil for both English and Spanish.

mejiaj commented on web-components/src/components/usa-banner/index.js:
```diff
@@ -92,32 +240,15 @@ export class UsaBanner extends LitElement {
+                      </slot> </strong
+                    ><br />
+                    <slot name="https-text">
+                      ${unsafeHTML(https.text1)} (${this.svgLock()})
+                      ${unsafeHTML(https.text2)} .${tld} ${https.text3}
```
> [!NOTE]  
> Correctly renders the `<strong>` HTML tags in labels defined in data.

mahoneycm commented on web-components/package.json:
```diff
@@ -6,7 +6,9 @@
     "build": "vite build",
-    "preview": "vite preview"
+    "preview": "vite preview",
+    "prettier:js": "npx prettier --check './src/components/**/*.js'",
+    "prettier:js:fix": "npx prettier --write './src/components/**/*.js'"
```
Is this colon syntax the standard for package scripts? Looks a bit odd imo. Curious if we should consider kebab-case or camelCase in the new repo 🤔 

Any opinions here?

heymatthenry replied:
@mahoneycm I think it's a pretty common convention to use the colon for quasi-namespacing, particularly in projects that use `run-all`.

mahoneycm commented on web-components/src/components/usa-banner/index.js:
```diff
@@ -76,9 +218,15 @@ export class UsaBanner extends LitElement {
-                    <strong>Official websites use .gov</strong><br />A
-                    <strong>.gov</strong> website belongs to an official
-                    government organization in the United States.
+                    <strong>
+                      <slot name="domain-heading">
+                        ${domain.heading} .${tld}
+                      </slot>
+                    </strong>
+                    <br />
+                    <slot name="domain-text">
+                      ${domain.text1} <strong>.${tld}</strong> ${domain.text2}
+                    </slot>
```
Do you think there'd be befit in breaking any of these sections out into their own templates? 

I don't know if it'd be over abstracted in a sense but we could break the inner content into a template and/or even the https and domain blocks into their own template functions to consolidate the HTML and make it more readable in the JS

mejiaj replied:
Added in b580613;  Not seeing a big benefit, but trying nonetheless because it could increase template readability.


The only small downside is now I have to pass the `TLD` to each template.

```js
${this.domainTemplate(tld)}
```

I'm sure there's a cleaner way to fallback to default value if user doesn't pass anything. Will look to iterate on this in future re-work for banner.

**Example**
```html
<usa-banner tld="banana"></usa-banner> 

<!-- Expecting `gov` because TLD of banana is invalid. -->
```

heymatthenry commented on web-components/src/components/usa-banner/index.js:
```diff
@@ -20,19 +31,146 @@ export class UsaBanner extends LitElement {
+  }
+
+
+  // ! CSS won't work if comments added inside css``.
```
If we're going to standardize on keeping styles separate, the (S)CSS should probably go in its own file. For `usa-link` I just moved the styles into a `usa-link.css.js`, but maybe here it might make sense to do `usa-banner.scss.js` to signal that there's still a SASS dependency here.

mejiaj replied:
Separated styles in 3549483.

> [!CAUTION]
> My only concern here is that a developer could see this and write SCSS instead of CSS, which is what we're moving towards.

Commit:
Run prettier.

Commit:
Create templates for domain and HTTPS.

Experimenting based on suggestion here: https://github.com/uswds/uswds-next/pull/12#discussion_r1635104164

Commit:
Separate styles.

Using `.scss.js` extension to signify there's still a SCSS dependency.

heymatthenry reviewed (approved):
Thanks for all your great work on this so far!

Commit:
Merge branch 'develop' of github.com:uswds/uswds-next into jm-lit-banner

Commit:
Reorg tasks and rename preview for clarity.

Commit:
Create banner story.

mejiaj commented:
@heymatthenry great point, I've added stories for all variants!

<details><summary>Screenshot</summary>
<p></p>

![image](https://github.com/uswds/uswds-next/assets/3385219/8a72ed96-3786-4e6b-a39d-fa2ce8724d35)
</details>
